From 516d030c9b7b10e4011504c4446a7f539f06da65 Mon Sep 17 00:00:00 2001
From: Paul Harrison <pharrison@mythtv.org>
Date: Tue, 8 Jul 2014 18:42:55 +0100
Subject: [PATCH] MythZoneMinder: Fix the idle and shutdown menu options not
 working in the menu

This passes any unhandled selections to the default TVMenuCallback handler.
---
 mythplugins/mythzoneminder/mythzoneminder/main.cpp |   34 ++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/mythplugins/mythzoneminder/mythzoneminder/main.cpp b/mythplugins/mythzoneminder/mythzoneminder/main.cpp
index a437470..edf34af 100644
--- a/mythplugins/mythzoneminder/mythzoneminder/main.cpp
+++ b/mythplugins/mythzoneminder/mythzoneminder/main.cpp
@@ -86,6 +86,10 @@ static void runZMEventView(void)
         mainStack->AddScreen(events);
 }
 
+// these point to the the mainmenu callback if found
+static void (*m_callback)(void *, QString &) = NULL;
+static void *m_callbackdata = NULL;
+
 static void ZoneMinderCallback(void *data, QString &selection)
 {
     (void) data;
@@ -98,16 +102,46 @@ static void ZoneMinderCallback(void *data, QString &selection)
         runZMLiveView();
     else if (sel == "zm_event_viewer")
         runZMEventView();
+    else
+    {
+        // if we have found the mainmenu callback
+        // pass the selection on to it
+        if (m_callback && m_callbackdata)
+            m_callback(m_callbackdata, selection);
+    }
 }
 
 static int runMenu(QString which_menu)
 {
     QString themedir = GetMythUI()->GetThemeDir();
 
+    // find the 'mainmenu' MythThemedMenu so we can use the callback from it
+    MythThemedMenu *mainMenu = NULL;
+    QObject *parentObject = GetMythMainWindow()->GetMainStack()->GetTopScreen();
+
+    while (parentObject)
+    {
+        mainMenu = dynamic_cast<MythThemedMenu *>(parentObject);
+
+        if (mainMenu && mainMenu->objectName() == "mainmenu")
+            break;
+
+        parentObject = parentObject->parent();
+    }
+
     MythThemedMenu *diag = new MythThemedMenu(
         themedir, which_menu, GetMythMainWindow()->GetMainStack(),
         "zoneminder menu");
 
+    // save the callback from the main menu
+    if (mainMenu)
+        mainMenu->getCallback(&m_callback, &m_callbackdata);
+    else
+    {
+        m_callback = NULL;
+        m_callbackdata = NULL;
+    }
+
     diag->setCallback(ZoneMinderCallback, NULL);
     diag->setKillable();
 
-- 
1.7.10.2

